testing/python/metafunc.py

```diff
@@ -1503,6 +1505,66 @@ def test_it(x): pass
         result = pytester.runpytest()
         assert result.ret == 0
 
+    def test_reordering_with_scopeless_and_just_indirect_parametrization(
+        self, pytester: Pytester
+    ) -> None:
+        pytester.makeconftest(
+            """
+            import pytest
+
+            @pytest.fixture(scope="package")
+            def fixture1():
+                pass
+            """
+        )
+        pytester.makepyfile(
+            """
+            import pytest
+
+            @pytest.fixture(scope="module")
+            def fixture0():
+                pass
+
+            @pytest.fixture(scope="module")
+            def fixture1(fixture0):
+                pass
+
+            @pytest.mark.parametrize("fixture1", [0], indirect=True)
+            def test_0(fixture1):
+                pass
+
+            @pytest.fixture(scope="module")
+            def fixture():
+                pass
+
+            @pytest.mark.parametrize("fixture", [0], indirect=True)
+            def test_1(fixture):
+                pass
+
+            def test_2():
+                pass
+
+            class Test:
+                @pytest.fixture(scope="class")
+                def fixture(self):
```

Comment:
I think this is important for testing this particular code path:

```suggestion
                def fixture(self, fixture):
```

This change might make test3 to be reordered before test2, in order to share to module-scoped `fixture` -- I'm not sure (maybe it will only happen once we reorder based on param value and not param index).

Reply:
I'll add it but it does not change the order since the only produced fixture key of `test_3` is `FixtureArgKey('fixture', 0, 'test_module', 'Test')` and the only produced fixture key of `test_1` is `FixtureArgKey('fixture', 0, 'test_module', None)`. Basing reordering on param value does not change it either as the module-scoped `fixture` is shadowed behind the class-scoped one for `test_3`. In other words we only pick `fixturedefs[-1]` to produce a key.

If you think we should consider shadowed-but-used fixtures into reordering as well, one way to that end is to check if `fixturedefs[-1] itself uses `fixture` or not so that we could produce a key for it as well. A general solution to considering shadowed-but-used fixtures was among the items in #11234 (`Considering used shadowed fixture dependencies for reordering as well. This was done by changing fixturemanager::getfixtureclosure algo from BFS to DFS.`) but I removed it as I thought it would not get accepted by the team. We could move forward with that as well.

Reply:
> I'll add it but it does not change the order since the only produced fixture key of test_3 is FixtureArgKey('fixture', 0, 'test_module', 'Test') and the only produced fixture key of test_1 is FixtureArgKey('fixture', 0, 'test_module', None)

Right, I forgot about the `item_cls`.

Regarding reordering based on shadowed-but-used fixtures, naively it makes sense to me -- they are used after all so why should they be ignored?

> but I removed it as I thought it would not get accepted by the team. We could move forward with that as well.

Perhaps after the current batch of changes we can look into it and think more deeply about it.